crates/ark/src/lsp/completions/completion_utils.rs

```diff
@@ -0,0 +1,123 @@
+use tree_sitter::Node;
+
+use crate::lsp::document_context::DocumentContext;
+use crate::lsp::traits::rope::RopeExt;
+use crate::treesitter::node_text;
+use crate::treesitter::NodeType;
+use crate::treesitter::NodeTypeExt;
+use crate::treesitter::UnaryOperatorType;
+
+#[allow(dead_code)]
+pub struct NodeContext<'a> {
+    pub node: Node<'a>,
+    pub node_text: String,
+    pub parent_node: Option<Node<'a>>,
+    pub parent_node_text: String,
+    pub grandparent_node: Option<Node<'a>>,
+    pub grandparent_node_text: String,
+    pub greatgrandparent_node: Option<Node<'a>>,
+    pub greatgrandparent_node_text: String,
+}
+
+pub fn gather_completion_context<'a>(context: &'a DocumentContext<'a>) -> NodeContext<'a> {
+    let mut node = context.node;
+    // trailing underscore to avoid conflict with the node_text function
+    let mut node_text_ = node_text(&node, &context.document.contents).unwrap_or_default();
+
+    let mut parent_node = None;
+    let mut parent_node_text = String::new();
+    let mut grandparent_node = None;
+    let mut grandparent_node_text = String::new();
+    let mut greatgrandparent_node = None;
+    let mut greatgrandparent_node_text = String::new();
+
+    if let Some(mut parent) = node.parent() {
+        // if we are completing "thi" as part of "pkgname::thi", the node we want to
+        // start walking up the AST from is the parent node
+        if parent.is_namespace_operator() {
+            node = parent;
+            node_text_ = node_text(&node, &context.document.contents).unwrap_or_default();
+            parent = node.parent().unwrap();
+        }
+
+        parent_node_text = node_text(&parent, &context.document.contents).unwrap_or_default();
+        parent_node = Some(parent);
+
+        if let Some(grandparent) = parent.parent() {
+            grandparent_node_text =
+                node_text(&grandparent, &context.document.contents).unwrap_or_default();
+            grandparent_node = Some(grandparent);
+
+            if let Some(greatgrandparent) = grandparent.parent() {
+                greatgrandparent_node_text =
+                    node_text(&greatgrandparent, &context.document.contents).unwrap_or_default();
+                greatgrandparent_node = Some(greatgrandparent);
+            }
+        }
+    }
+
+    NodeContext {
+        node,
+        node_text: node_text_,
+        parent_node,
+        parent_node_text,
+        grandparent_node,
+        grandparent_node_text,
+        greatgrandparent_node,
+        greatgrandparent_node_text,
+    }
+}
+
+pub fn check_for_function_value(context: &DocumentContext, node_context: &NodeContext) -> bool {
```

Comment:
I revised these to be less nested -- I liked that aspect of #369 once I discovered it.